fhevm/operators_comparison_gas.go

```diff
@@ -141,3 +142,63 @@ func fheIfThenElseRequiredGas(environment EVMEnvironment, input []byte) uint64 {
 	}
 	return environment.FhevmParams().GasCosts.FheIfThenElse[second.fheUintType()]
 }
+
+func fheArrayEqRequiredGas(environment EVMEnvironment, input []byte) uint64 {
+	logger := environment.GetLogger()
+
+	unpacked, err := arrayEqMethod.Inputs.UnpackValues(input)
+	if err != nil {
+		msg := "fheArrayEqRun RequiredGas() failed to unpack input"
+		logger.Error(msg, "err", err)
+		return 0
+	}
+
+	if len(unpacked) != 2 {
+		err := fmt.Errorf("fheArrayEqRun RequiredGas() unexpected unpacked len: %d", len(unpacked))
+		logger.Error(err.Error())
+		return 0
+	}
+
+	lhs, err := getVerifiedCiphertexts(environment, unpacked[0])
+	if err != nil {
+		msg := "fheArrayEqRun RequiredGas() failed to get lhs to verified ciphertexts"
+		logger.Error(msg, "err", err)
+		return 0
+	}
+
+	rhs, err := getVerifiedCiphertexts(environment, unpacked[1])
+	if err != nil {
+		msg := "fheArrayEqRun RequiredGas() failed to get rhs to verified ciphertexts"
+		logger.Error(msg, "err", err)
+		return 0
+	}
+
+	if len(lhs) != len(rhs) || (len(lhs) == 0 && len(rhs) == 0) {
+		return environment.FhevmParams().GasCosts.FheTrivialEncrypt[tfhe.FheBool]
+	}
+
+	numElements := len(lhs)
+	elementType := lhs[0].Type()
+	// TODO: tie to supported types in tfhe.TfheCiphertext.EqArray()
+	if elementType != tfhe.FheUint4 && elementType != tfhe.FheUint8 && elementType != tfhe.FheUint16 && elementType != tfhe.FheUint32 && elementType != tfhe.FheUint64 {
+		return 0
+	}
+	for i := range lhs {
+		if lhs[i].Type() != elementType || rhs[i].Type() != elementType {
+			return 0
+		}
+	}
+
+	numBits := elementType.NumBits() * uint(numElements)
+	if numBits <= 16 {
+		return environment.FhevmParams().GasCosts.FheEq[tfhe.FheUint16]
+	} else if numBits <= 32 {
+		return environment.FhevmParams().GasCosts.FheEq[tfhe.FheUint32]
+	} else if numBits <= 64 {
+		return environment.FhevmParams().GasCosts.FheEq[tfhe.FheUint64]
+	} else if numBits <= 160 {
```

Comment:
Idea is that the gas cost for up to 160 bits of array comparison is the one for FheUint160, because the runtime is mostly dependent on the bit count and not on the datatype. Meaning that comparing two arrays of 160 bits each (irrespective of data type) is the same as comparing two FheUint160 integers.

However, for number of bits > 160, we don't have a benchmark and I came up with some number (`FheArrayEqBigArrayFactor`) to add to the gas cost, but that is totally made up. That leads to:

```return (environment.FhevmParams().GasCosts.FheEq[tfhe.FheUint160] + environment.FhevmParams().GasCosts.FheArrayEqBigArrayFactor) * (uint64(numBits) / 160)```.

Runtime (and hence gas) is supposed to be logarithmic (but not log2) on the number of bits, but I didn't implement such a calculation.

Any idea on whether we should be more precise at that point?

Reply:
I don't think we need precision, we just need to be sure to not under estimate.